frontend/src/lib/fabricCanvasUtils.ts

```diff
@@ -0,0 +1,364 @@
+    strokeWidth: strokeWidth,
+    selectable: true,
+  });
+  canvas?.add(line);
+  canvas?.setActiveObject(line);
+  return false;
+};
+
+export const addRectangle = (canvas: fabric.Canvas | null, options?: { fillColor?: string; width?: number; height?: number }): void => {
+  const { fillColor = "red", width = 200, height = 100 } = options || {};
+  const rect = new fabric.Rect({
+    left: 100,
+    top: 100,
+    fill: fillColor,
+    width: width,
+    height: height,
+    angle: 0,
+    transparentCorners: false,
+  });
+  canvas?.add(rect);
+  canvas?.setActiveObject(rect);
+};
+
+export const addCircle = (canvas: fabric.Canvas | null, options?: { fillColor?: string; radius?: number }): void => {
+  const { fillColor = "green", radius = 50 } = options || {};
+  const circle = new fabric.Circle({
+    radius: radius,
+    fill: fillColor,
+    left: 200,
+    top: 200,
+  });
+  canvas?.add(circle);
+  canvas?.setActiveObject(circle);
+};
+
+const removeImagesFromCanvas = (canvas:fabric.Canvas | null) => {
+  const imageObjects = canvas?.getObjects('image') as fabric.Image[];
+  const imagesData = imageObjects.map((imageObj) => ({
+    src: imageObj.getSrc(),
+    left: imageObj.left,
+    top: imageObj.top,
+    scaleX: imageObj.scaleX,
+    scaleY: imageObj.scaleY,
+    angle: imageObj.angle,
+  }));
+  imageObjects.forEach((imageObj) => canvas?.remove(imageObj));
+  canvas?.renderAll();
+  return imagesData;
+};
+
+const restoreImagesToCanvas = (canvas:fabric.Canvas | null, imagesData:any) => {
+  imagesData.forEach((imgData:any) => {
+    fabric.Image.fromURL(imgData.src, (img) => {
+      img.set({
+        left: imgData.left,
+        top: imgData.top,
+        scaleX: imgData.scaleX,
+        scaleY: imgData.scaleY,
+        angle: imgData.angle,
+      });
+      canvas?.add(img);
+      canvas?.renderAll();
+    });
+  });
+};
+
+export const exportToPDF = async (canvas: fabric.Canvas | null): Promise<void> => {
+  if (!canvas) return;
+
+  const imagesData = removeImagesFromCanvas(canvas);
+
+  const imgData = canvas.toDataURL({
+    format: 'png',
+    quality: 1,
+  });
+
+  const pdf = new jsPDF({
+    orientation: 'landscape',
+    unit: 'px',
+    format: [canvas.getWidth(), canvas.getHeight()],
+  });
+
+  pdf.addImage(imgData, 'PNG', 0, 0, canvas.getWidth(), canvas.getHeight());
+  pdf.save('canvas.pdf');
+
+  restoreImagesToCanvas(canvas, imagesData);
+};
+
+
+
+
+
+export const addImage = (canvas: fabric.Canvas | null, imageUrl: string, options?: { scaleX?: number; scaleY?: number; left?: number; top?: number }): void => {
+  const { scaleX = 1, scaleY = 1, left = 0, top = 0 } = options || {};
+  fabric.Image.fromURL(imageUrl, (img) => {
+    img.set({
+      scaleX: scaleX,
+      scaleY: scaleY,
+      left: left,
+      top: top,
+    });
+    canvas?.add(img);
+  });
+
+};
+
+export const resizeImage = (imageSrc: string, maxWidth: number, maxHeight: number, callback: (resizedImage: string) => void) => {
```

Comment:
I don't understand what is happening here or why it's here 😢  And why are we creating a new canva?

Reply:
it was needed for adding new images

Reply:
To clarify, I wanted to scale images that are too large, and this unusual method was allowing us to do it.

Reply:
Oh, I see. I guess the method's name wasn't clear enough for me. The method is not for any resizing, but specifically for 'fitting' the image into a smaller area if it's too big, right? The image gets shrunk so it can fit. If so, what about `fitImageByShrinking` or `shrinkAndFitImage`?

Reply:
okay you are right